### ISSUE TITLE
The integer negative zero formats as "-0.0"

### ISSUE THREAD

**Issue by zathras** [dart-lang/sdk#56292]
### Steps to reproduce
Use "flutter run" to run the four-line program provided below in Chrome, and observe that the int formats as "-0.0".

### Expected results
```
A Dart VM Service on Chrome is available at: http://127.0.0.1:59786/bfAu4QGpcN4=
 The int "-0" formats like this:  0
```
### Actual results
```
A Dart VM Service on Chrome is available at: http://127.0.0.1:59786/bfAu4QGpcN4=
 The int "-0" formats like this:  -0.0
```
### Code sample

`cat lib/main.dart`
```
void main() {
  int i = -0;
  print(' The int "-0" formats like this:  $i');
}
```

### Shows up as null in debugger!

This is amusing.  The "`int`" -0 shows up as `null` in Android Studio:

![Screenshot 2024-07-19 at 6 21 06 AM](https://github.com/user-attachments/assets/878b6c80-1563-4b5c-b9e4-913952e7f6d2)

### Flutter Doctor output
```
Doctor output
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 3.22.3, on macOS 14.5 23F79 darwin-arm64, locale
    en-US)
[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
[✓] Xcode - develop for iOS and macOS (Xcode 15.3)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2024.1)
[✓] Connected device (3 available)
[✓] Network resources

• No issues found!
```

### Other references

Originally filed as https://github.com/flutter/flutter/issues/152008 .
This bug was noticed in an application here:  https://github.com/flutter/flutter/issues/152008

**Comment by julemand101** [comment]
Can you add to the issue which platforms you are observing this issue? E.g. web, native (and which native platforms).?

**Comment by lrhn** [comment]
When compiling for the web, dart integers are represented by JavaScript numbers, which are doubles.
That's why `-0` exists at all, it doesn't in native integers.

The JS runtime has no way to know whether you *wanted* the value `-0.0` or got it accidentally by writing `-0` on an int.
It might be able to guess from the static type, but when reaching `(-0).toString()`, it doesn't know where it came from.

So, this is working as intended for web numbers.

(Can't say what the debugger thing means.)

**Comment by zathras** [comment]
Yes, I know about the limitations of the underlying JavaScript runtime.

That of course doesn't mean those limitations can't be patched over, like Dart does in so many other areas.  Having an `int` that `isNegative` but is also `== 0` is at the very least a violation of the principle of least astonishment, and it is certainly possible to patch over JavaScript's brokenness in this case.  If Dart has a well-specified formal language definition in this area, it could well violate that, too.

You've successfully explained *why* the bug exists.  It's still fair to say that it *is* a bug - no sane person would expect something declared as an `int` in a purportedly type-safe language to format as "-0.0", and otherwise behave in decidedly non-`int`-like ways.  Whether or not this bug is considered to be important enough to fix is, of course, another matter.

As an aside, I'm sure everybody hopes JS can be jettisoned in favor of WASM some day, so all the JS brain-damage can go away.  But the power of legacy is fierce.

**Comment by nshahan** [comment]
@zathras 
There is a document describing the differences in number representations on various platforms here:
https://dart.dev/guides/language/numbers

**Comment by zathras** [comment]
> There is a document describing the differences in number representations on various platforms here:
> https://dart.dev/guides/language/numbers

Indeed there is.  No where does it suggest that this (unexpected) behavior will leak through the abstraction.  Indeed, quite the contrary:

> Depending on the platform, those numeric types have different, **hidden** implementations

(emphasis mine)

**Comment by sigmundch** [comment]
> ... It's still fair to say that it is a bug - no sane person would expect ...

I assume you mean well, just want to call out this sentence comes across a bit strong.

> ... No where does it suggest that this (unexpected) behavior will leak through the abstraction.

Please note the article actually goes in great detail about [differences in behavior](https://dart.dev/guides/language/numbers#differences-in-behavior) to help everyone navigate those tricky corner cases. It so happens that the section about the [string representation](https://dart.dev/guides/language/numbers#string-representation) highlights this example of `-0.0` as one of those surprising corner cases.

An important takeaway is that Dart has always allowed platform-specific representations and semantics for numbers, and (even though it appears as a bug) sometimes it is an intentional choice to improve performance, code size, or interoperability. The tradeoff is sometimes more obvious for other things discussed in the article (like bit-wise operators or identity), but they all come from the same goal.

It so happens that `-0.0` is one of those cases that has really bugged us (no pun intended) too and we've revisited more than others. It even behaves inconsistently with some optimizations (see https://github.com/dart-lang/sdk/issues/49870). @rakudrama may have additional comments to share here.

Re: wasm - good news is that our dart2wasm backend does match the VM semantics and will behave as you expect in this case :tada:

**Comment by zathras** [comment]
So an `int` can be JSON encoded (on the web ruintime) such that, when it is decoded on a normal runtime, it comes back as a `double`.  But only in the case of -0.  Not for e.g. -1, 0, or 1.  That's an...  interesting choice.

I guess the intent was to preserve the distinction between -0 and 0 for numbers that are intended to be doubles?  It's certainly not the thing about "deferring to JS," because that's not what JS does.  I see how both choices (following JS, or making the special case for -0.0 that was done) are bad in different ways, but hoo boy, changing the type of the value emitted as JSON for an `int` to a double in this one odd corner case is inviting very subtle bugs!

Yeah, of course WASM fixes this.  It supports a real `int` type, after all, avoiding all kinds of other breakages.
